Consider repl NOT healthy with IO_Thread connecting and last attempt failed

mattlord opened the pull request:
## Description
[We considered](https://github.com/vitessio/vitess/blob/6c20177eaff5e739d6eeee5b69b6112891e8bba6/go/mysql/flavor.go#L316-L321) the `Slave_IO_Running` state of `Connecting` as equivalent to `Running` in the [replication status](https://dev.mysql.com/doc/refman/en/replication-administration-status.html) results that we get from MySQL. I'm assuming this was done to avoid flapping on low traffic systems due to the [`-slave_net_timeout`](https://dev.mysql.com/doc/refman/en/replication-options-replica.html#sysvar_slave_net_timeout) reconnects and avoid doing a tablet repair when replication was healthy (with this PR we can distinguish running from healthy states and take differing actions depending on what we want).

After https://github.com/vitessio/vitess/pull/9308 we properly estimate the replica lag when MySQL is telling us that it does not know, meaning it returns a `NULL` value for the `Seconds_Behind_Master` field. But in some cases — e.g. it appears to happen when attempting the *first* connection to the source — MySQL will report a `Seconds_Behind_Master` value of `0` (meaning fully caught up and no lag) even when it is not connected to its replication source and has _failed to reconnect_ — meaning that this is not a simple reconnect for any reason (e.g. [`-slave_net_timeout`](https://dev.mysql.com/doc/refman/en/replication-options-replica.html#sysvar_slave_net_timeout)), but a reconnect with one or more _failures/errors_. This PR handles this latter case within Vitess by continuing to treat `Connecting` as equivalent to `Running` via `ReplicationStatus.Healthy()` — to prevent the noted flapping and errant/unnecessary tablet repairs — **_unless_**  we had an IO error the last time we tried to reconnect to the replication source. 

> Note: I think that [this](https://github.com/vitessio/vitess/pull/9789/files#r820941187) was also what in effect caused the bug seen in https://github.com/vitessio/vitess/issues/9788 as the new replica tablet was considered healthy when in fact it had not been able to connect to its source (for the very first time)

## Related Issue(s)
- Fixes: https://github.com/vitessio/vitess/issues/9788
- Related to: https://github.com/vitessio/vitess/pull/9308

## Checklist
- [x] Should this PR be backported? NO
- [x] Tests are not required (I don't think)
- [x] Documentation is not required

shlomi-noach commented:
I have a vague memory of having this discussion elsewhere in the code. Also, I had a similar in `orchestrator`. IMO the correct way to handle this is to avoid using a `bool` for `Slave_IO_Running`. The problem with keeping the behavior as before this PR is clear. But this PR also creates a confusion: it will say the IO thread is not running when it could be in the interim of completing a connection. So that's also confusing. Example for a risk to this approach: you check replication status, you get `false` for IO thread running; you assume replication is stopped; you run some logic that relies on this assumption, but unknown to you, the IO thread finally succeeds completing a connection and all of a sudden replication is running.

So IMO the state of IO thread should be an enum like `stopped`, `connecting`, `running`.

mattlord commented:
> I have a vague memory of having this discussion elsewhere in the code. Also, I had a similar in `orchestrator`. IMO the correct way to handle this is to avoid using a `bool` for `Slave_IO_Running`. The problem with keeping the behavior as before this PR is clear. But this PR also creates a confusion: it will say the IO thread is not running when it could be in the interim of completing a connection. So that's also confusing. Example for a risk to this approach: you check replication status, you get `false` for IO thread running; you assume replication is stopped; you run some logic that relies on this assumption, but unknown to you, the IO thread finally succeeds completing a connection and all of a sudden replication is running.
> 
> So IMO the state of IO thread should be an enum like `stopped`, `connecting`, `running`.

I like this idea. It's more accurate (the SQL_Thread really does only have a binary state whereas the IO_Thread has 3 states) and it allows the caller to decide whether or not to consider connecting and running equivalent or perform different actions. I'll work on this now.

Thank you!

shlomi-noach commented on go/mysql/constants.go:
```diff
@@ -677,3 +678,33 @@ func IsSchemaApplyError(err error) bool {
+// off,no,false == ReplicationStateStopped
+// connecting == ReplicationStateConnecting
+// anything else == ReplicationStateUnknown
+func ReplicationStatusToState(s interface{}) ReplicationState {
```
IMHO the function should accept a `string`; it doesn't make sense for it to accept any type. It should be the caller's duty to provide a `string`.

mattlord replied:
I don't feel strongly about it. I thought that sending a boolean would be nicer for some existing tests and uses. I can remove it though if there's good reason to.

mattlord replied:
I'll do this now just to simplify things. Simplicity is itself a good reason on second thought. Thanks!

shlomi-noach commented on go/mysql/constants.go:
```diff
@@ -677,3 +678,33 @@ func IsSchemaApplyError(err error) bool {
+		return ReplicationStateConnecting
+	} else {
+		return ReplicationStateUnknown
+	}
```
purely a styling matter, but I think `switch` is cleaner here:
```go
	switch strings.ToLower(s) {
	case "yes", "on", "true":
		return ReplicationStateRunning
	case "no", "off", "false":
		return ReplicationStateStopped
	case "connecting":
		return ReplicationStateConnecting
	default:
		return ReplicationStateUnknown
	}
```
feel free to ignore any styling comment of mine.

mattlord replied:
Sure, I don't have a strong preference either way. I started on this patch and then switched and don't remember why. 🙂 Was making too many changes while tired, I think. I'll move to this. Thanks!

shlomi-noach commented on go/mysql/flavor_mysqlgr.go:
```diff
@@ -141,32 +141,32 @@ func (mysqlGRFlavor) status(c *Conn) (ReplicationStatus, error) {
 		FROM performance_schema.replication_connection_status
 		WHERE CHANNEL_NAME='%s'`, chanel)
-	var ioThreadRunning bool
+	var connectionState ReplicationState
```
the default will be `ReplicationStateRunning`

mattlord replied:
Good catch! 😄 Will change the ordering so that 0 is unknown as you noted. Thanks!

shlomi-noach commented on go/mysql/constants.go:
```diff
@@ -677,3 +678,33 @@ func IsSchemaApplyError(err error) bool {
+	ReplicationStateRunning ReplicationState = iota
+	ReplicationStateStopped
+	ReplicationStateConnecting
+	ReplicationStateUnknown
```
let's make `ReplicationStateUnknown` the first - hence default - value. See below

mattlord commented:
@shlomi-noach I applied all of your great suggestions and the tests are now passing. If you have time for another quick review pass (not urgent), I'd most appreciate it. 

Thank you again!

shlomi-noach reviewed (approved):
:rocket:

mattlord commented on go/vt/mysqlctl/replication.go:
```diff
@@ -47,7 +47,7 @@ func WaitForReplicationStart(mysqld MysqlDaemon, replicaStartDeadline int) error
 		}
 
-		if status.ReplicationRunning() {
+		if status.Running() {
```
Note: this was a bug. It should be using `Running()` and NOT `Healthy()` because it's supposed to wait until we know that we can connect to the source.

GuptaManan100 reviewed (commented):
Everything looks good to me! I just have 1 question. Is there any way to cleanup the old fields in Proto. I know they cannot be removed immediately for upgrade-downgrade considerations, but can they be marked for deprecation or something like that?

shlomi-noach reviewed (approved):
LGTM. I don';t have full context into the problem, and am assuming that the existing tests cover the behavior exhibited by this change.

mattlord commented:
> Everything looks good to me! I just have 1 question. Is there any way to cleanup the old fields in Proto. I know they cannot be removed immediately for upgrade-downgrade considerations, but can they be marked for deprecation or something like that?

Good question! I was wondering that too, but I don't know the answer. @deepthi do we have a defined process or an example to follow regarding making backwards compatibility breaking protobuf changes?

It would be nice to e.g. remove the `io_thread_running` and `io_thread_connecting` booleans and replace them with an `io_state` int field.

deepthi commented:
> > Everything looks good to me! I just have 1 question. Is there any way to cleanup the old fields in Proto. I know they cannot be removed immediately for upgrade-downgrade considerations, but can they be marked for deprecation or something like that?
> 
> Good question! I was wondering that too, but I don't know the answer. @deepthi do we have a defined process or an example to follow regarding making backwards compatibility breaking protobuf changes?
> 
> It would be nice to e.g. remove the `io_thread_running` and `io_thread_connecting` booleans and replace them with an `io_state` int field.

The way to make such changes to proto definitions is to use `reserved`. you rename the old fields to `reserved` and add the new field. I realize that we have a couple of fields with `deprecated=true`, those slipped through reviews, we should never have done that.

mattlord commented:
> > > Everything looks good to me! I just have 1 question. Is there any way to cleanup the old fields in Proto. I know they cannot be removed immediately for upgrade-downgrade considerations, but can they be marked for deprecation or something like that?
> > 
> > 
> > Good question! I was wondering that too, but I don't know the answer. @deepthi do we have a defined process or an example to follow regarding making backwards compatibility breaking protobuf changes?
> > It would be nice to e.g. remove the `io_thread_running` and `io_thread_connecting` booleans and replace them with an `io_state` int field.
> 
> The way to make such changes to proto definitions is to use `reserved`. you replace the old field with something like `reserved 2` and add the new field. I realize that we have a couple of fields with `deprecated=true`, those slipped through reviews, we should never have done that.

OK, thanks! I'll investigate to see if there's a way I can safely effectively deprecate the `io_thread_running` and `sql_thread_connecting` booleans and effectively replace them with `io_state` and `sql_state` ints (and get rid of the `io_thread_connecting` boolean added in the PR). I was assuming that would be forbidden but perhaps I was wrong. If we can do that it would keep the proto to struct translation clean and simple.

mattlord commented:
I apologize, but I'm going to rebase via FF merge and force-push as GitHub Actions is somehow checking out an invalid snapshot and I'm getting CI errors for code that *doesn't exist* in my branch's `HEAD`. I'll share another comment with the diff covering the changes since the last reviews once the tests are passing again.

mattlord commented:
Ugh, still happening. You can see an example here with errors about code that doesn't exist in my brach: https://github.com/vitessio/vitess/runs/5482587430?check_suite_focus=true

mattlord commented:
Force pushing from a fixed local branch to this PRs remote didn't seem to do it either. 😢 Moving this to a new PR: https://github.com/vitessio/vitess/pull/9853
